### DISCUSSION TITLE
Cameras support

### DISCUSSION THREAD

**Discussion by kfir-o** [NVlabs/instant-ngp#546] [Q&A]
I'm trying to run the instant-ngp over dataset which contains images from 5 different cameras (mounted together side by side).
each camera has its unique parameters (FocalLength, SensorSize, PrincipalPointX, PrincipalPointY, K1, K2, K3, P1, P2).
The best fit seems to be OpenCVFisheyeCameraModel with K4=0 but I have doubt if it's supported by the entire pipeline and I only need to edit the colmap2nerf script.

I'm not sure how to bind those parameters (and others, like camera Latitude, Longitude, AbsAlt, RltAlt, Pitch, Roll and Yaw at each image) correctly since the NeRF's transforms.json format is not documented.

Any help \ advice will be much appreciated (:

**Comment by ataraxyz** [comment]
Hey! I would like to add an upvote to this. I have classic Photogrammetry datasets. With drone footage AND high res cameras. Due to the fact that the camera model is set to a single one in colmap this fails. Editing colmap2nerf.py:89 to use multiple cameras ( --ImageReader.single_camera 0 ) lets me generate the transforms, but the resulting nerf model is way less accurate then the model from a single camera source. Surely this is technically possible? Can you help?
THANK YOU FOR YOUR WORK!

**Comment by Tom94** [comment]
Hi there, the camera matrices in `transforms.json` are standard matrices in homogeneous coordinates, describing the position and orientation of the cameras.

The OpenCV fisheye model is indeed not supported, although we do support the general regular OpenCV camera model, the parameters of which can be supplied in `transforms.json` per image; see https://github.com/NVlabs/instant-ngp/discussions/797 for an example.

**Reply by ataraxyz** [reply]
Hey! Thank you for that, 
my main issue has nothing to do with the Camera model, all I am trying is mixing a drone with Canon footage on a single dataset, And as per above had to change the single camera parameter. The canon solve is not adding anything relevant to the NeRF set. 

As an update, I did try and just use the Canon images with not the best results either. Can you just confirm that one camera only yields the best ( and per the current state of the code only possible ) results?

**Reply by Tom94** [reply]
As a general rule, when multiple poses show the same content from similar angles, the worst of the images will "pollute" the reconstruction. This is an unfortunate side-effect of NeRF and, while it can be remedied with LoD tricks (like mip-NeRF), this is not done in instant-ngp.

So: using images from multiple cameras _can_ help, but only if they show the object from sufficiently different viewpoints.

**Comment by ishipachev** [comment]
Hi @Tom94,

May I ask you to point out the right sections of code where this fisheye camera support has to be integrated? 
I would like to see how complex the issue and maybe try to bring the support of fisheye cameras there if I see it is doable by my knowledge and experience.
I know that NN also backpropagates an error into camera params, so the whole thing maybe really complex at the end.
